apps/core/scenes/bookingDetail/sectorDetails/segments/SegmentMap.js

```diff
@@ -6,50 +6,117 @@ import { StyleSheet } from '@kiwicom/universal-components';
 import { graphql, createFragmentContainer } from '@kiwicom/margarita-relay';
 import { getCenter } from 'geolib';
 import memoize from 'memoize-one';
-import stringify from 'json-stable-stringify';
+import { uniqBy, prop } from 'ramda';
 
 import type { SegmentMap as BookingType } from './__generated__/SegmentMap.graphql';
 import MapLines from './MapLines';
+import SegmentMapMarker from './SegmentMapMarker';
 
 type Props = {|
   +data: ?BookingType,
 |};
 
+type RouteStop = {|
+  +stop: ?{|
+    +city: ?{|
+      +name: ?string,
+    |},
+    +locationId: ?string,
+    +coordinates: ?{|
+      +latitude: ?number,
+      +longitude: ?number,
+    |},
+  |},
+|};
+
+type Segments = ?$ReadOnlyArray<?{|
+  +arrival: ?RouteStop,
+  +departure: ?RouteStop,
+|}>;
+
+type MarkerData = {|
+  +latitude: ?number,
+  +longitude: ?number,
+  +name: ?string,
+  +locationId: string,
+|};
+
 const DEFAULT_PADDING = { top: 40, right: 40, bottom: 40, left: 40 };
 
-const mapCoordinates = memoize(
-  (segmentLocations: ?$PropertyType<BookingType, 'segmentLocations'>) => {
-    const coordinates = segmentLocations ?? [];
-    return coordinates.map(coordinate => ({
-      latitude: coordinate?.lat,
-      longitude: coordinate?.lng,
-    }));
-  },
-);
+const mapMarkerData = memoize((props: Props) => {
+  switch (props.data?.type) {
+    case 'BOOKING_ONE_WAY':
+      return mapSegmentsToMarkerData(props.data?.sector?.segments);
+    case 'BOOKING_RETURN':
+      return mapSegmentsToMarkerData(getReturnSegments(props));
+    case 'BOOKING_MULTICITY':
+      return mapSegmentsToMarkerData(getMulticitySegments(props));
+    default:
+      return null;
+  }
+});
+
+const getReturnSegments = (props: Props) => {
+  const outbound = props.data?.outbound?.segments ?? [];
+  const inbound = props.data?.inbound?.segments ?? [];
+  return [...outbound, ...inbound];
+};
+
+const getMulticitySegments = (props: Props) => {
+  const sectors = props.data?.sectors ?? [];
+  return sectors.reduce((acc, curr) => {
+    const segments = curr?.segments ?? [];
+    return [...acc, ...segments];
+  }, []);
+};
 
-class SegmentMap extends React.Component<Props> {
+const mapSegmentsToMarkerData = (segments: Segments) => {
+  if (segments == null) {
+    return [];
+  }
+  const bookingSegments = segments.reduce((acc, curr) => {
+    return [
+      ...acc,
+      {
+        ...curr?.departure?.stop?.coordinates,
+        name: curr?.departure?.stop?.city?.name,
+        locationId: curr?.departure?.stop?.locationId ?? '',
+      },
+      {
+        ...curr?.arrival?.stop?.coordinates,
+        name: curr?.arrival?.stop?.city?.name,
+        locationId: curr?.arrival?.stop?.locationId ?? '',
+      },
+    ];
+  }, []);
+
+  return uniqBy<MarkerData, { +locationId: string }>(
```

Comment:
I find this difficult to read. Would it be possible to extract the types?

Reply:
🚀